pkg/services/zulip/zulip.go

```diff
@@ -0,0 +1,87 @@
+package zulip
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/containrrr/shoutrrr/pkg/services/standard"
+	"github.com/containrrr/shoutrrr/pkg/types"
+)
+
+// Service sends notifications to a pre-configured channel or user
+type Service struct {
+	standard.Standard
+	config *Config
+}
+
+const (
+	contentMaxSize = 10000 // bytes
+	topicMaxLength = 60    // characters
+)
+
+// Send a notification message to Zulip
+func (service *Service) Send(message string, params *types.Params) error {
+	config := *service.config
+
+	if params != nil {
+		if stream, found := (*params)["stream"]; found {
+			config.Stream = stream
```

Comment:
Aren't you mutating the config here? 

That would mean that if you run this two times, the first with params and the second one without, both would use the first ones values.

Reply:
I tried to avoid that by dereferencing the pointer to the config in https://github.com/containrrr/shoutrrr/pull/39/files#diff-b282a090227cdf789619987d9821e43fR27

Reply:
Well, you are dereferencing the pointer, which means that if `config` changes, no side effects leave the method. But changing `config.*` is not the same thing. 
The problem is that no copy of the structure has taken place, which means that changing the struct fields changes the shared memory.
I might be wrong, and/or juggling too many languages, though 😁, perhaps a test for it is in place regardless?

Reply:
I might be wrong, and/or juggling too many languages as well :smile:

I tried it and it worked as I expected.

But since this is not obvious (and we both get confused by switching between languages) I have made an explicit Clone() method on the Config struct and added tests to make sure it's actually a clone and not the same struct:
```patch
diff --git a/pkg/services/zulip/zulip.go b/pkg/services/zulip/zulip.go
index f4e016a..6366138 100644
--- a/pkg/services/zulip/zulip.go
+++ b/pkg/services/zulip/zulip.go
@@ -24,7 +24,10 @@ const (
 
 // Send a notification message to Zulip
 func (service *Service) Send(message string, params *types.Params) error {
-	config := *service.config
+	// Clone the config because we might modify stream and/or
+	// topic with values from the parameters and they should only
+	// change this Send().
+	config := service.config.Clone()
 
 	if params != nil {
 		if stream, found := (*params)["stream"]; found {
@@ -48,7 +51,7 @@ func (service *Service) Send(message string, params *types.Params) error {
 		return fmt.Errorf("message exceeds max size (%d bytes): was %d bytes", contentMaxSize, messageSize)
 	}
 
-	return service.doSend(&config, message)
+	return service.doSend(config, message)
 }
 
 // Initialize loads ServiceConfig from configURL and sets logger for this Service
diff --git a/pkg/services/zulip/zulip_config.go b/pkg/services/zulip/zulip_config.go
index d138343..ff9cb97 100644
--- a/pkg/services/zulip/zulip_config.go
+++ b/pkg/services/zulip/zulip_config.go
@@ -65,6 +65,18 @@ func (config *Config) SetURL(serviceURL *url.URL) error {
 	return nil
 }
 
+// Clone the config to a new Config struct
+func (config *Config) Clone() *Config {
+	return &Config{
+		BotMail: config.BotMail,
+		BotKey:  config.BotKey,
+		Host:    config.Host,
+		Path:    config.Path,
+		Stream:  config.Stream,
+		Topic:   config.Topic,
+	}
+}
+
 const (
 	// Scheme is the identifying part of this service's configuration URL
 	Scheme = "zulip"
diff --git a/pkg/services/zulip/zulip_test.go b/pkg/services/zulip/zulip_test.go
index 22939a3..65f6762 100644
--- a/pkg/services/zulip/zulip_test.go
+++ b/pkg/services/zulip/zulip_test.go
@@ -79,6 +79,36 @@ var _ = Describe("the zulip service", func() {
 		})
 	})
 	Describe("the zulip config", func() {
+		When("cloning a config object", func() {
+			It("the clone should have equal values", func() {
+				config1 := &zulip.Config{
+					BotMail: "bot-name@zulipchat.com",
+					BotKey:  "correcthorsebatterystable",
+					Host:    "example.zulipchat.com",
+					Path:    "api/v1/messages",
+					Stream:  "foo",
+					Topic:   "bar",
+				}
+
+				config2 := config1.Clone()
+
+				Expect(config1).To(Equal(config2))
+			})
+			It("the clone should not be the same struct", func() {
+				config1 := &zulip.Config{
+					BotMail: "bot-name@zulipchat.com",
+					BotKey:  "correcthorsebatterystable",
+					Host:    "example.zulipchat.com",
+					Path:    "api/v1/messages",
+					Stream:  "foo",
+					Topic:   "bar",
+				}
+
+				config2 := config1.Clone()
+
+				Expect(config1).NotTo(BeIdenticalTo(config2))
+			})
+		})
 		When("generating a config object", func() {
 			It("should generate a correct config object", func() {
 				zulipURL, err := url.Parse("zulip://bot-name%40zulipchat.com:correcthorsebatterystable@example.zulipchat.com?stream=foo&topic=bar")
```

Reply:
Yeah, I was wrong. When doing a short variable declaration of a struct, you are actually making a copy of that struct. Which does make sense, since it's fields are actually part of the variable, and not pointers. So you would only need to do the manual cloning when the struct itself contains references, like slices for example.

Reply:
Well, it _is_ confusing. I think being more explicit as it ended up being is way better so thanks for wondering :+1: